spi-cp2130.c

```diff
@@ -831,47 +831,37 @@ static int cp2130_spi_transfer_one_message(struct spi_master *master,
 
                 if (xfer->tx_buf && xfer->rx_buf) {
 			urb[CP2130_BULK_OFFSET_CMD] = CP2130_CMD_WRITEREAD;
-			/* usb write */
-			ret = usb_bulk_msg(dev->udev, xmit_pipe, urb,
-					   CP2130_BULK_OFFSET_DATA + xfer->len,
-					   &len, 200);
-			dev_dbg(&master->dev, "usb write %d", ret);
-			if (ret)
-				goto err;
-			/* usb read */
-			ret = usb_bulk_msg(dev->udev, recv_pipe,
-					   xfer->rx_buf, xfer->len,
-					   &len, 200);
-			dev_dbg(&master->dev, "usb read %d", ret);
-			if (ret)
-				goto err;
 		} else if (!xfer->rx_buf) {
 			/* prepare URB and submit sync */
 			urb[CP2130_BULK_OFFSET_CMD] = CP2130_CMD_WRITE;
-			/* usb write */
-			ret = usb_bulk_msg(dev->udev, xmit_pipe, urb,
-					   CP2130_BULK_OFFSET_DATA + xfer->len,
-					   &len, 200);
-			if (ret)
-				goto err;
 		} else if (!xfer->tx_buf) {
 			/* prepare URB and submit sync */
 			urb[CP2130_BULK_OFFSET_CMD] = CP2130_CMD_READ;
-			/* usb write */
-			ret = usb_bulk_msg(dev->udev, xmit_pipe, urb,
-					   CP2130_BULK_OFFSET_DATA,
-					   &len, 200);
-			if (ret)
-				goto err;
-			/* usb read */
-			ret = usb_bulk_msg(dev->udev, recv_pipe, xfer->rx_buf,
-					   xfer->len, &len, 200);
-			if (ret)
-				goto err;
 		}
 
-		udelay(xfer->delay_usecs);
-		mesg->actual_length += xfer->len;
+		/* usb write */
```

Comment:
There is no loop around the `usb_bulk_msg()` call on the `xmit_pipe`.  If the TX payload is > 56 bytes, then won't `usb_bulk_msg` fail to send all of the data?

Reply:
The best explanation for splitting the transmission I've found is this:

"Note that it is possible for a driver to send amounts of data to an endpoint that is bigger than this value, but the data will be divided up into wMaxPacketSize chunks when actually transmitted to the device"
https://www.safaribooksonline.com/library/view/linux-device-drivers/0596005903/ch13.html

From my tests if I try to transfer more than 64 bytes, some packets would stay buffered and the driver would print this message every second: "failed to read gpios"
With the loop, cp2130_spi_transfer_one_message would handle these received packets, return the right information to the application and no error message is printed by the irq worker code.

Reply:
Thanks for the explanation.  You have reminded me that I meant to read that chapter of ldd3.

Reply:
I think it will be necessary to split the writes into chunks of 64byte packets as well at least for the readwrite case as the cp2130 will not cache more than 64bytes of response, so for readwrite we need to write+read in chunks of 64byte.